libzfs_core/_libzfs_core.py

```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+            if size == 0:
+                break
+            data_bytes = os.read(fd, size)
+            result = {}
+            with nvlist_out(result) as nvp:
+                ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+            if ret != 0:
+                raise ZFSError(ret, "Failed to unpack list data")
+            entry = result['name']
+            if entry == name:
+                break
+            else:
+                entry = None
+    finally:
+        os.close(other_fd)
+        os.close(fd)
+
+    if entry is None:
+        raise exceptions.DatasetNotFound(name)
+    result = result['properties']
+    # In most cases the source of the property is uninteresting and the
+    # value alone is sufficient.  One exception is the 'mountpoint'
+    # property the final value of which is not the same as the inherited
+    # value.
+    mountpoint = result.get('mountpoint')
+    if mountpoint is not None:
+        mountpoint_src = mountpoint['source']
+        mountpoint_val = mountpoint['value']
+        # 'source' is the name of the dataset that has 'mountpoint' set
+        # to a non-default value and from which the current dataset inherits
+        # the property.  'source' can be the current dataset if its
+        # 'mountpoint' is explicitly set.
+        # 'source' can also be a special value like '$recvd', that case
+        # is equivalent to the property being set on the current dataset.
+        if not mountpoint_src.startswith('$'):
+            mountpoint_val = mountpoint_val + name[len(mountpoint_src):]
+    else:
+        mountpoint_val = '/' + name
+    result = { k: v['value'] for k, v in result.iteritems() }
+    if result.has_key('clones'):
+        result['clones'] = [ x for x in result['clones'] ]
+    result['mountpoint'] = mountpoint_val
+    return result
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
+    '''
+    List the children of the ZFS dataset.
+
+    :param str name: the name of the dataset.
+    :return: an iterator that produces the names of the children.
+    '''
+    (fd, other_fd) = lzc_list(name, {'recurse': 1, 'type': {'filesystem': None, 'volume': None}})
+
+    def _iterator():
+        try:
+            while True:
+                record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+                if not record_bytes:
+                    break
+                (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+                xlate.lzc_list_children_xlate_error(err, name)
+                if size == 0:
+                    break
+                data_bytes = os.read(fd, size)
+                result = {}
+                with nvlist_out(result) as nvp:
+                    ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+                if ret != 0:
+                    raise ZFSError(ret, "Failed to unpack list data")
+                child = result['name']
+                if child != name:
+                    yield child
+        finally:
```

Comment:
`finally` inside a generator is a somewhat uncertain thing.  Either the generator must be iterated to exhaustion (presumably the caller) or the generator must be collected by the garbage collector.  If one of these things happens, this finally block will run.  If the generator ends up referenced from an uncollectable cycle, the finally block will never run.

It's very difficult to ensure objects don't end up in uncollectable cycles (particularly when writing code for someone else to use).  I tend to look at this feature of generators as being basically unusable.

One alternative would be to replace this generator with a normal function that collects results into a list and then returns the list.  This has the downside of greater memory use.  But it is at least straightforward.

The constant factor on memory usage could be reduced by just reading all the bytes and then using a generator to parse them and produce results.  The Python objects are probably much larger than the byte-stream representation of the same data so maybe this would even be a big win - not sure, probably easily measured.

Another option might be to make `lzc_list` return a ... context manager ... maybe?

```py
with lzc_list(name, options) as listing:
    os.read(listing.read_fd, ...)
    ...
```

and have the two file descriptors closed by the context manager on exit.  Though if you go on to use this in a generator ...

```py
def lzc_list_children(..):
    with lzc_list(name, options) as listing:
        stuff = os.read(listing.read_fd, ...)
        ...
        yield ...
```

Then actually you're probably back to the same old problem.  So nevermind that, I suppose.

Perhaps the sensible thing is to go with the memory-hungry solution and hope there's an optimization to be made later (either with or without an interface change...)